Add Prospect notebook

weaverba137 opened the pull request:
This notebook demonstrates obtaining SDSS spectra from the spectrum service and displaying it with Prospect.

*Important note:* I developed this on the user-facing, production notebook server, which does not yet have prospect installed, so I still have to load prospect from a checkout.  This will be a one-line change when prospect is finally installed.

rnikutta reviewed (commented):
* Suggest using "Astro Data Lab" in lieu of "Data Lab" where possible (e.g. NB title, Goals, Summary)
* In the Imports cell, please add `from getpass import getpass`
* In Authentication: "require a login (i.e. you will need a registered user account." --> closing parens is missing
* Forgive my ignorance here: "All spPlate files have a common wavelength solution, so a spPlate file can be represented by a Spectrum1D object." --> Is it "all spPlate files have a common wavelength solution", or "all spectra from the same plate have a common wavelength solution"? (or is each spectrum a seperate spPlate file, i.e. the two statements are equivalent?)
* "so a spPlate file can be represented by a Spectrum1D object. This object is an input to the visualization system, so we need to get outputs from the spectrum service that match the inputs to the visualization." --> I don't understand the last sentence. What do you mean by "the outputs from the service [must] match the inputs to the viz"?
* "SDSS plates are circular, so we'll do a cone search around the nominal plate center. This gets us the spectrum identifiers we need to retrieve the spectra themselves." --> I thought `spec.query()` could fetch all IDs from a single plate given as an argument. But it doesn't seem so. Was this option removed? (or maybe I just mis-remember it..)
* First code cell below "Redshift and other Metadata" --> Not questioning Prospect, but curious why it expects Z_ERR, and all field names in capitals?
* Same issue: "we need to reformat it to match the expectations of Prospect" --> If these are documented somewhere, maybe add a link to these Prospect definitions?
* The NB currently does:
```
sdss_ids = spec.query(r[0], r[1], 1.5, fmt='array', out='', constraint='plate=2955 ORDER BY specobjid LIMIT 50')
```
followed by:
```
q = "SELECT specobjid, class, subclass, z, zerr, rchi2diff, zwarning FROM sdss_dr16.specobj WHERE specobjid IN ({0}) ORDER BY specobjid".format(','.join(map(str, sdss_ids.tolist())))
sdss_z = qc.query(sql=q, fmt='table')
```
But you could get the same sdss_z table with just:
```
sdss_z = spec.query(r[0], r[1], 1.5, fields='specobjid,class,subclass,z,zerr,rchi2diff,zwarning', constraint='plate=2955 ORDER BY specobjid LIMIT 50')
```
* This very long cell has no intro or post-execution info; it's quite opaque to understand what's being done:
```
q = """SELECT s.specobjid, s.targetobjid, s.ra, s.dec,
  s.primtarget, s.sectarget,
  s.boss_target1,
  s.ancillary_target1, s.ancillary_target2,
  s.eboss_target0, s.eboss_target1, s.eboss_target2,
  p.u, p.g, p.r, p.i, p.z
FROM sdss_dr16.specobj AS s
LEFT JOIN sdss_dr16.photoplate AS p
ON s.bestobjid = p.objid
WHERE specobjid IN ({0})
ORDER BY specobjid""".format(','.join(map(str, sdss_ids.tolist())))
sdss_plugmap = qc.query(sql=q, fmt='table')
for c in ('ra', 'dec', 'primtarget', 'sectarget',
          'boss_target1',
          'ancillary_target1', 'ancillary_target2',
          'eboss_target0', 'eboss_target1', 'eboss_target2'):
    sdss_plugmap.rename_column(c, c.upper())
sdss_plugmap['OBJID'] = np.vstack((np.bitwise_and(sdss_plugmap['targetobjid'] >> 32, 2**16 - 1).data,
                                   np.bitwise_and(sdss_plugmap['targetobjid'] >> 48, 2**11 - 1).data,
                                   np.bitwise_and(sdss_plugmap['targetobjid'] >> 29, 2**3 - 1).data,
                                   np.bitwise_and(sdss_plugmap['targetobjid'] >> 16, 2**12 - 1).data,
                                   np.bitwise_and(sdss_plugmap['targetobjid'], 2**16 - 1).data)).T
sdss_plugmap['MAG'] = np.vstack((np.where(np.isnan(sdss_plugmap['u'].data), 0, sdss_plugmap['u'].data),
                                 np.where(np.isnan(sdss_plugmap['g'].data), 0, sdss_plugmap['g'].data),
                                 np.where(np.isnan(sdss_plugmap['r'].data), 0, sdss_plugmap['r'].data),
                                 np.where(np.isnan(sdss_plugmap['i'].data), 0, sdss_plugmap['i'].data),
                                 np.where(np.isnan(sdss_plugmap['z'].data), 0, sdss_plugmap['z'].data))).T

sdss.meta['plugmap'] = sdss_plugmap
sdss_plugmap
```
* The code cell under "Launch the Visualization" has many unexplained arguments. Not all are obvious; some (simple) explanatory text could help.
* After the Bokeh widget is loaded, it'd be good to add a few pointers about what one can now do. E.g. you can hover the mouse cursor over any part of the spectrum; you can zoom in/out (with mouse wheel or with zoom box), pan (click and drag), and what the (most important) widget controls do; otherwise the compexity of the viewer can be quite overwhelming, especially to novice Prospectors.
* The second code cell under "SDSS Spectra and Redshifts from Files" fails for me (this is on dltest):
```
%%time
sdss_file = read_spPlate(sc.get(sdss_spectra, mode='fileobj'), limit=50)
sdss_file

---------------------------------------------------------------------------
InvalidSchema                             Traceback (most recent call last)
<timed exec> in <module>

/data0/sw/anaconda3/lib/python3.7/site-packages/noaodatalab-2.19.0-py3.7.egg/dl/Util.py in __call__(self, *args, **kw)
     80             return function(self.obj, *args, **kw)
     81         else:
---> 82             return function(*args, **kw)
     83 
     84     def __repr__(self):

/data0/sw/anaconda3/lib/python3.7/site-packages/noaodatalab-2.19.0-py3.7.egg/dl/storeClient.py in get(optval, fr, to, token, mode, verbose, debug, timeout)
    410         return sc_client._get(fr=optval, to=to, token=def_token(token),
    411                             mode=mode, verbose=verbose, debug=debug,
--> 412                             timeout=timeout)
    413 
    414 @multimethod('sc',0,False)

/data0/sw/anaconda3/lib/python3.7/site-packages/noaodatalab-2.19.0-py3.7.egg/dl/storeClient.py in _get(self, token, fr, to, mode, verbose, debug, timeout)
   1600             url = requests.get(self.svc_url + "/get?name=%s" % nm,
   1601                                headers=hdrs)
-> 1602             r = requests.get(url.text, stream=False, headers=hdrs)
   1603             if mode == 'text':
   1604                 return scToString(r.content)

/data0/sw/anaconda3/lib/python3.7/site-packages/requests/api.py in get(url, params, **kwargs)
     73 
     74     kwargs.setdefault('allow_redirects', True)
---> 75     return request('get', url, params=params, **kwargs)
     76 
     77 

/data0/sw/anaconda3/lib/python3.7/site-packages/requests/api.py in request(method, url, **kwargs)
     58     # cases, and look like a memory leak in others.
     59     with sessions.Session() as session:
---> 60         return session.request(method=method, url=url, **kwargs)
     61 
     62 

/data0/sw/anaconda3/lib/python3.7/site-packages/requests/sessions.py in request(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)
    531         }
    532         send_kwargs.update(settings)
--> 533         resp = self.send(prep, **send_kwargs)
    534 
    535         return resp

/data0/sw/anaconda3/lib/python3.7/site-packages/requests/sessions.py in send(self, request, **kwargs)
    638 
    639         # Get the appropriate adapter to use
--> 640         adapter = self.get_adapter(url=request.url)
    641 
    642         # Start time (approximately) of the request

/data0/sw/anaconda3/lib/python3.7/site-packages/requests/sessions.py in get_adapter(self, url)
    729 
    730         # Nothing matches :-/
--> 731         raise InvalidSchema("No connection adapters were found for '%s'" % url)
    732 
    733     def close(self):

InvalidSchema: No connection adapters were found for 'String index out of range: -1'
```
* With that cell failing, I can test the two cells under it, as they rely on the result.

weaverba137 commented:
@rnikutta, thank you for the comments.  I'm working on these now, please read on:

> * Suggest using "Astro Data Lab" in lieu of "Data Lab" where possible (e.g. NB title, Goals, Summary)
> * In the Imports cell, please add `from getpass import getpass`
> * In Authentication: "require a login (i.e. you will need a registered user account." --> closing parens is missing

Done

> * Forgive my ignorance here: "All spPlate files have a common wavelength solution, so a spPlate file can be represented by a Spectrum1D object." --> Is it "all spPlate files have a common wavelength solution", or "all spectra from the same plate have a common wavelength solution"? (or is each spectrum a seperate spPlate file, i.e. the two statements are equivalent?)

Sorry for the confusion.  The correct phrasing should be "all spectra in an *individual* spPlate file have a common wavelength solution."  I have updated the text.

> * "so a spPlate file can be represented by a Spectrum1D object. This object is an input to the visualization system, so we need to get outputs from the spectrum service that match the inputs to the visualization." --> I don't understand the last sentence. What do you mean by "the outputs from the service [must] match the inputs to the viz"?

That phrasing is left over from a time when this notebook was extracting spectra in a very different object representation.  It should say "the Spectrum1D object is a format accepted by the prospect visualization package".

> * "SDSS plates are circular, so we'll do a cone search around the nominal plate center. This gets us the spectrum identifiers we need to retrieve the spectra themselves." --> I thought `spec.query()` could fetch all IDs from a single plate given as an argument. But it doesn't seem so. Was this option removed? (or maybe I just mis-remember it..)

From the documentation I've seen `spec.query()` does require some form of coordinates as input.  @mjfitzpatrick , can you comment here?  There are examples of using `qc.query(sql='...', ...)`, but not `spec.query(sql='...', ...)`.

> * First code cell below "Redshift and other Metadata" --> Not questioning Prospect, but curious why it expects Z_ERR, and all field names in capitals?
> * Same issue: "we need to reformat it to match the expectations of Prospect" --> If these are documented somewhere, maybe add a link to these Prospect definitions?

The spectral metadata provided by SDSS has these fields capitalized.  Prospect expects the input data to match that.  This is also important in certain cases for distinguishing SDSS metadata from DESI metadata.  I can add more explanatory text, but there is not much else I can do here.

> * The NB currently does: ...

Well, I thought from all the discussion of specClient that finding the spectrum IDs is a critical step.  So by having that separate, that step is highlighted.  This also clearly separates the very important step of finding the IDs from the step of finding the metadata.  I could be convinced otherwise, but I have a slight preference for that separation of functionality.

> * This very long cell has no intro or post-execution info; it's quite opaque to understand what's being done:
> * The code cell under "Launch the Visualization" has many unexplained arguments. Not all are obvious; some (simple) explanatory text could help.

I will add more explanation for these.

> * After the Bokeh widget is loaded, it'd be good to add a few pointers about what one can now do....

OK, but would you prefer that text *above* or *below* the Bokeh cell?  On many displays, especially laptop-size, the Bokeh widget will take up the entire display, so do you prefer scrolling up or scrolling down for that?

> * The second code cell under "SDSS Spectra and Redshifts from Files" fails for me (this is on dltest):

This works for me on gp12/production.  I notice that `requests` is involved.  Wasn't there an issue with that on dltest?

mjfitzpatrick commented:
On Thu, Feb 11, 2021 at 9:11 AM Benjamin Alan Weaver <
notifications@github.com> wrote:

>
>    - "SDSS plates are circular, so we'll do a cone search around the
>    nominal plate center. This gets us the spectrum identifiers we need to
>    retrieve the spectra themselves." --> I thought spec.query() could
>    fetch all IDs from a single plate given as an argument. But it doesn't seem
>    so. Was this option removed? (or maybe I just mis-remember it..)
>
> From the documentation I've seen spec.query() does require some form of
> coordinates as input. @mjfitzpatrick <https://github.com/mjfitzpatrick> ,
> can you comment here? There are examples of using qc.query(sql='...', ...),
> but not spec.query(sql='...', ...).
>
You're confusing query() and getSpec():  In getSpec() you can pass in a
template identifier like "[(1963,54331)]" as a plate/mjd tuple and it will
expand the fibers automatically.  For query(), you would need to use the
constraint query with no positional arguments, e.g.
"query(constraint='plateid = 1963 and mjd = 54331')".

Note also there is a difference between queryClient.query() and
spec.query() as relates to specobjid:  The latter will automatically cast
specobjid to a uint64 whereas the qc.query() returns a possibly signed
integer as stored in the DB.

-Mike

rnikutta commented:
Thanks @weaverba137 

> > * The NB currently does: ...
> 
> Well, I thought from all the discussion of specClient that finding the spectrum IDs is a critical step. So by having that separate, that step is highlighted. This also clearly separates the very important step of finding the IDs from the step of finding the metadata. I could be convinced otherwise, but I have a slight preference for that separation of functionality.

Understood, makes sense.

> > * After the Bokeh widget is loaded, it'd be good to add a few pointers about what one can now do....
> 
> OK, but would you prefer that text _above_ or _below_ the Bokeh cell? On many displays, especially laptop-size, the Bokeh widget will take up the entire display, so do you prefer scrolling up or scrolling down for that?

Maybe just a sentence above (something with "We will initialize the spectral viewer. It allows browse through spectra, has zoom and pan capabilities, and other functions."

And under the widget, one could explain a bit more.

All other points you and Mike have clarified, thank you.

weaverba137 commented:
Notebook is updated, please review.

rnikutta commented:
Prospect is a really cool tool!

Here my 2nd round:

* Update version string in first cell before PR is merged
* Clean up (remove)
```
# What the heck? Can't use a semicolon any longer?
# q = "SELECT ra, dec FROM sdss_dr16.platex WHERE plate = %d;" % plate
```
* "Here we find spectra corresponding to SDSS plate 2955, but using the spectrum service instead of reading the spectra from a file."
--> But the query cell just below finds RA and Dec of a plate, not spectra.
* Cell with:
```sdss_ids = spec.query(r[0], r[1], 1.5, fmt='array', out='', constraint='plate=2955 ORDER BY specobjid LIMIT 50')```
--> Why hard-code 2995 here? It has been defined in the previous section as the 'plate' variable.
* Cell with:
```sdss = spec.getSpec(sdss_ids, fmt='Spectrum1D', align=True)```
--> Say one sentence what 'align=True' does?
* "In the cells below, we wil extract ..." --> will
*  "this changes the numerical scale factor on the flux unit not the flux unit itself"
--> add comma: "... on the flux unit, not the flux unit itself"
* "but will change in future vresions of prospect"
--> versions, Prospect
* "Spectrum slider (< / >): Use these buttons or the slider page through the loaded spectra."
--> "to page through"
* Cell with:
``` os.environ['SPECTRO_REDUX'] = ...```
 --> Why the choice to set an env variable, rather than a simple NB variable? (it's only used in this very cell)
* Could not review the final cells, from cell
```
%%time
sdss_file = read_spPlate(sc.get(sdss_spectra, mode='fileobj'), limit=50)
sdss_file
```
because it throws an error (please see Slack #notebook channel).

weaverba137 commented:
Updated after second round of comments.  *Note*: the HTML file is out of sync with the notebook.  Remind me to update that before merging.

> Update version string in first cell before PR is merged

Done

> * Clean up (remove)
> ```python
> # What the heck? Can't use a semicolon any longer?
> # q = "SELECT ra, dec FROM sdss_dr16.platex WHERE plate = %d;" % plate
> ```

I still can't use a semicolon in a SQL query on production.  Why is that not fixed?  Note that this is `queryClient.query()` not `specClient.query()` in this cell.

> "Here we find spectra corresponding to SDSS plate 2955, but using the spectrum service instead of reading the spectra from a file."

Rephrased.

> Why hard-code 2995 here? It has been defined in the previous section as the 'plate' variable.

Agreed, fixed.

> Say one sentence what 'align=True' does?

Done.

> "In the cells below, we wil extract ..." --> will

Done.

> add comma:

Done.

> "but will change in future vresions of prospect"

Done.

> to page through

Done.

> Why the choice to set an env variable, rather than a simple NB variable? (it's only used in this very cell)

Hmmm, historical reasons; not needed by Prospect.  Changed to a simple variable.

> Could not review the final cells, from cell

Still works for me on production, so must be an issue with dltest.

mjfitzpatrick commented:
On Fri, Feb 19, 2021 at 12:11 PM Benjamin Alan Weaver <
notifications@github.com> wrote:

> I still can't use a semicolon in a SQL query on production. Why is that
> not fixed?
>
 It is already fixed, Seems to be pending a PR merge and deployment....

weaverba137 commented:
@mjfitzpatrick, thanx, I thought that had already been done.

rnikutta commented:
Just pinged second reviewer.

rnikutta commented:
I did a full test run now, looks good. Only three things from my side to update before merging:
* add back in semicolon in query (once that capability has been deployed at DL production)
* Remove the related comment from the NB
* Update the HTML render as well

Pinging @stephjuneau to please test as well, as 2nd reviewer.

weaverba137 commented:
This should be ready for final check and merge now.

rnikutta commented:
Thank you! Will test & merge this afternoon (after a bunch of meetings)

rnikutta commented:
Hi @weaverba137 I tested the NB on prod and it looks good and runs well. Would you mind doing just one more commit to this PR? The rendered HTML has no outputs. But the HTML is best for a quick preview of a NB, so it would be good to have the output in the rendered HTML as well. Thanks!

weaverba137 commented:
OK, how much do you want rendered?  I don't know now a bokeh plot will render to an HTML file.  In principle it is mostly JavaScript, but it's very complex.  And there are actually two places where bokeh is used, the specClient and the equivalent test using files.

rnikutta commented:
I think it's fine if the Prospect widget  is not rendered. It's static HTML after all. Can you give it a try?

weaverba137 commented:
OK, done.

rnikutta commented:
Merged. Many thanks!
